### PR TITLE
Email Notice when Export Download is Ready

### PR DESCRIPTION
**Pull request by xbryanc**
Added a button that emails the user when his/her export is ready to download. Once the user has clicked the button and the request has been sent to the server, the user may close the window since the rest is taken care of server-side.

**This was done using a VM, and absolute_reverse caused a hang when clicking the button. This still must be tested on a computer with sufficient memory @dannyroberts **

Edit: Verified to work on a regular computer

### CODE REVIEW

File: corehq/apps/export/static/export/js/download_export.ng.js [hunk 1]
```diff
@@ -286,6 +287,21 @@
...
             hqImport('analytix/js/kissmetrix').track.event("Clicked Download button");
         };
+
+        $scope.sendEmailUponCompletion = function () {
+            setTimeout(function () {
+                if ($scope.download_id !== undefined) {
+                    $.ajax({
+                        method: 'POST',
+                        dataType: 'json',
+                        url: '../../../../../add_export_email_request/',
```
Comment (orangejenny) [inline corehq/apps/export/static/export/js/download_export.ng.js]: There's infrastructure for providing URLs to JavaScript without needing to hard code them. Check out https://github.com/dimagi/js-guide/blob/master/integration-patterns.md#django-urls

File: corehq/apps/export/migrations/0004_emailrequest_email_address.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+    dependencies = [
+        ('export', '0003_emailrequest'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='emailrequest',
+            name='email_address',
+            field=models.CharField(default='foo@gmail.com', max_length=255),
```
Comment (orangejenny) [inline corehq/apps/export/migrations/0004_emailrequest_email_address.py]: Odd default - why not blank string?

File: corehq/apps/export/templates/export/partial/export_download_progress.html [hunk 3]
```diff
@@ -52,8 +52,19 @@
                                 <span ng-show="!isMultimediaDownload">{% trans "Collecting Data" %}</span>
                                 <span ng-show="isMultimediaDownload">{% trans "Collecting Multimedia" %}</span>
                         </button>
+                        {% if request|toggle_enabled:"EMAIL_EXPORT_WHEN_DONE_BUTTON" %}
+                        <button type="button"
+                                class="btn btn-primary btn-full-width"
```
Comment (orangejenny) [inline corehq/apps/export/templates/export/partial/export_download_progress.html]: `btn-primary` is only for a single primary action on a page. I'd switch this to `btn-default`.

File: corehq/ex-submodules/soil/util.py [hunk 4]
```diff
@@ -88,6 +92,20 @@ def get_download_context(download_id, message=None, require_result=False):
...
 
 
+def process_email_request(download_id, email_address):
+    dropbox_url = absolute_reverse('dropbox_upload', args=(download_id,))
+    try:
+        allow_dropbox_sync = get_download_context(download_id).get('allow_dropbox_sync', False)
+    except TaskFailedError:
+        allow_dropbox_sync = False
+    download_url = "{}?get_file".format(absolute_reverse('retrieve_download', args=(download_id,)))
+    email_body = 'Your CommCare export is ready! Click on the link below to download your ' + \
```
Comment (orangejenny) [inline corehq/ex-submodules/soil/util.py]: translate?

Comment (orangejenny) [reply corehq/ex-submodules/soil/util.py]: same with the subject line below

File: corehq/apps/export/views.py [hunk 5]
```diff
@@ -2481,6 +2482,23 @@ def can_download_daily_saved_export(export, domain, couch_user):
     return False
 
 
+@login_and_domain_required
```
Comment (orangejenny) [inline corehq/apps/export/views.py]: Might as well also use `@require_POST`

File: corehq/apps/export/views.py [hunk 6]
```diff
@@ -2481,6 +2482,23 @@ def can_download_daily_saved_export(export, domain, couch_user):
...
+def add_export_email_request(request, domain):
+    download_id = request.POST.get('download_id')
+    email_address = request.couch_user.get_email()
+    try:
+        download_context = get_download_context(download_id)
+    except TaskFailedError:
+        return HttpResponseServerError()
+    if download_id is None or email_address is None:
+        return HttpResponseBadRequest()
```
Comment (orangejenny) [inline corehq/apps/export/views.py]: Error message would be nice.

File: corehq/apps/export/templates/export/partial/export_download_progress.html [hunk 7]
```diff
@@ -52,8 +52,19 @@
...
+                                ng-init="sendEmailFlag = false"
+                                ng-disabled="sendEmailFlag"
+                                ng-click="sendEmailUponCompletion(); sendEmailFlag = true;">
+                                {% trans "Email me when this is done" %}
+                                <i class="fa fa-check" ng-show="sendEmailFlag"></i>
+                        </button>
+                        {% endif %}
                         {% if show_sync_to_dropbox %}
-                        <button button="submit"
+                        <button type="submit"
```
Comment (orangejenny) [inline corehq/apps/export/templates/export/partial/export_download_progress.html]: Curious, why was this change necessary?

**Review by orangejenny — COMMENTED** [review]
Didn't read closely, just skimmed for best practices.

Seems like a good feature in general, great for those with low connectivity. @dannyroberts is there a plan to get it un-feature-flagged?

**Comment by orangejenny** [comment]
> This was done using a VM, and absolute_reverse caused a hang when clicking the button.

If responsiveness is a concern - and we probably have users who have machines with as little memory as your VM - it'd be good to use [disableButton](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/hqwebapp/static/hqwebapp/js/hq.helpers.js#L138-L141) which'll add a spinner to the button so the user doesn't think the page is just not responding.

File: corehq/apps/export/static/export/js/download_export.ng.js (hunk 1 above)
Comment (dannyroberts) [reply corehq/apps/export/static/export/js/download_export.ng.js]: +1 to this. @xbryanc I know I showed you how to do this so you could get over the hump, but definitely worth fixing to have this follow that practice. (Keeps links from breaking if we change URLs, etc., but also is just standard practice that helps others follow the code.)

File: corehq/apps/export/templates/export/partial/export_download_progress.html (hunk 7 above)
Comment (dannyroberts) [reply corehq/apps/export/templates/export/partial/export_download_progress.html]: This confused us, but is just a random change. Is `<button button="submit">` normal/valid html? I assumed it was a typo.

File: corehq/ex-submodules/soil/util.py (hunk 4 above)
Comment (dannyroberts) [reply corehq/ex-submodules/soil/util.py]: @xbryanc "translating" here means marking for translation. Every bit of text we show the user should be marked for translation, so that when a user who speaks French sees it, they see it in french. (In this particular case, there may be a bit more we need to do to make this user's language preference available at this point in the code, but we can handle that down the road—at least with it marked for translation our translators can translate it in the meantime.)

Marking for translation is explained here: https://docs.djangoproject.com/en/1.11/topics/i18n/translation/. In addition, the use of the `+` operator for string concatenation is strongly discouraged in python; use `'My favorite food is {}!'.format('bananas')`-style string formatting instead.

File: corehq/apps/export/models/new.py [hunk 8]
```diff
@@ -2675,6 +2675,12 @@ def delete(self, using=None, keep_parents=False):
         return super(DataFile, self).delete(using, keep_parents)
 
 
+class EmailRequest(models.Model):
+    domain = models.CharField(max_length=255)
+    download_id = models.CharField(max_length=255)
+    email_address = models.CharField(max_length=255)
```
Comment (dannyroberts) [inline corehq/apps/export/models/new.py]: While we may only need an email address here, I think it would be better to put the user's `user_id` and then use that to look up the user and finally get the email address once you're in the task. There are a couple reasons for this. First, the user is the richer context from which you can derive the `email_address`—if later on we need to know other things about the user (like their language preference for instance) it will be nice to have available.

If you don't change this now, that's not the end of the world, but if it's not too hard, I think it'd be a minor improvement.

File: corehq/apps/export/static/export/js/download_export.ng.js [hunk 9]
```diff
@@ -286,6 +287,21 @@
                             hqImport('export/js/utils').capitalize(exportDownloadService.exportType), "Saved");
             hqImport('analytix/js/kissmetrix').track.event("Clicked Download button");
         };
+
+        $scope.sendEmailUponCompletion = function () {
+            setTimeout(function () {
```
Comment (dannyroberts) [inline corehq/apps/export/static/export/js/download_export.ng.js]: I would add a comment here about why you're doing this in an asynchronous loop. It may not be clear to the reader that it's to wait until `download_id` is available in the scope.

**Review by dannyroberts — COMMENTED** [review]
Looking good. Left a few comments

**Comment by orangejenny** [comment]
@xbryanc Danny gently pointed out that my comment about our users' machines makes no sense, feel free to ignore.

File: corehq/apps/export/templates/export/partial/export_download_progress.html (hunk 7 above)
Comment (orangejenny) [reply corehq/apps/export/templates/export/partial/export_download_progress.html]: The original code does look like a typo, but I had thought that the default button type was `button` and switching to `submit` could change behavior. But I just looked it up and learned the [default is submit](https://stackoverflow.com/questions/31644723/what-is-the-default-button-type), so yeah, should be a fine change.

Comment (dannyroberts) [reply corehq/apps/export/templates/export/partial/export_download_progress.html]: Ha! Thanks for digging. That's kinda funny, but makes sense. This worked for the person who wrote it but not for the reason that they thought it did.

File: corehq/ex-submodules/soil/util.py [hunk 10]
```diff
@@ -94,16 +95,18 @@ def get_download_context(download_id, message=None, require_result=False):
...
-    email_body = 'Your CommCare export is ready! Click on the link below to download your ' + \
-                 'requested data:\n' + download_url
-    dropbox_message = '\n\nYou can also upload your data to Dropbox with the link below:\n' + \
-                      dropbox_url if allow_dropbox_sync else ''
-    send_HTML_email('CommCare Export Complete', email_address, email_body + dropbox_message)
+    dropbox_message = ''
+    if allow_dropbox_sync:
+        dropbox_message = ('\n\nYou can also upload your data to Dropbox with the link below:\n'
+                           '{}').format(dropbox_url)
+    email_body = ('Your CommCare export is ready! Click on the link below to download your requested data:\n'
```
Comment (dannyroberts) [inline corehq/ex-submodules/soil/util.py]: I think you didn't actually translate these? You need to use `_` the way you did below.

**Review by dannyroberts — APPROVED** [review]
